torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1848,8 +1850,13 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   work->numelIn_ = inputs[0].numel();
   work->numelOut_ = outputs[0].numel();
 
+  // Notify graphs before we check the capture status preemptively
+  at::cuda::CUDAGraph::inc_pending_event_queries();
+
   if (!coalescing_state_ && capture_status == c10::cuda::CaptureStatus::None) {
     workEnqueue(work);
+  } else {
+    at::cuda::CUDAGraph::dec_pending_event_queries();
```

Comment:
I suggest to increment before `workEnqueue` and decrement after the work is completed.

Reply:
I thought about this, but we must preemptively increment here as incrementing after the capture status is checked is potentially unsafe. In this case, we can have an interaction where the observed capture status here is `None` but at the same time the graphs code begins a capture (updating capture status), but it is now too late to notify graphs code that there is work enqueued. By preemptively incrementing, we notify the graphs code that there many potentially be work enqueued until we are certain that either 1. the work is done, or 2. we don't actually enqueue it as we have received notification that a capture has begun.

Reply:
I see, now it makes sense. Thanks for the explanation! :+1: